src/panels/lovelace/cards/hui-shopping-list-card.ts

```diff
@@ -0,0 +1,236 @@
+    this._items = [];
+    this._fetchData();
+  }
+
+  public connectedCallback() {
+    super.connectedCallback();
+
+    if (this.hass) {
+      this.hass.connection
+        .subscribeEvents(() => this._fetchData(), "shopping_list_updated")
+        .then(
+          (unsub: () => Promise<void>): void => {
+            this._unsubEvents = unsub;
+          }
+        );
+      this._fetchData();
+    }
+  }
+
+  public disconnectedCallback() {
+    super.disconnectedCallback();
+
+    if (this._unsubEvents) {
+      this._unsubEvents();
+    }
+  }
+
+  protected render(): TemplateResult {
+    if (!this._config || !this.hass) {
+      return html``;
+    }
+
+    return html`
+      ${this.renderStyle()}
+      <ha-card .header="${this._config.title}">
+        <div class="addRow">
+          <paper-input
+            id='addBox'
+            placeholder='${this.localize("ui.card.shopping-list.add_item")}'
+            @keydown=${this._addKeyPress}
+            no-label-float
+          ></paper-input>
+          <paper-icon-button
+            slot="item-icon"
+            icon="hass:plus"
+            @click=${this._addItem}
+          ></paper-icon-button>
+          <paper-icon-button
+            slot="item-icon"
+            icon="hass:notification-clear-all"
+            @click=${this._clearItems}
+          ></paper-icon-button>
+        </div>
+
+      ${repeat(
+        this._items!,
+        (item) => html`
+        <div class="editRow">
+          <paper-checkbox
+            slot="item-icon"
+            ?checked=${item.complete}
+            .itemId=${item.id}
+            @click=${this._completeItem}
+            tabindex='0'
+          ></paper-checkbox>
+          <paper-item-body>
+            <paper-input
+              no-label-float
+              .value='${item.name}'
+              .itemId=${item.id}
+              @change=${this._saveEdit}
+            ></paper-input>
+          </paper-item-body>
+        </div>
+        `
+      )}
+      </ha-card>
+    `;
+  }
+
+  private renderStyle(): TemplateResult {
+    return html`
+      <style>
+        .addRow, .editRow {
+          display: flex;
+          flex-direction: row;
+        }
+        .addRow {
+          padding-left: 16px;
+        }
+        paper-icon-item {
+          border-top: 1px solid var(--divider-color);
+        }
+        paper-icon-item:first-child {
+          border-top: 0;
+        }
+        paper-checkbox {
+          padding: 11px;
+        }
+        paper-input {
+          --paper-input-container-underline: {
+            display: none;
+          }
+          --paper-input-container-underline-focus: {
+            display: none;
+          }
+          position: relative;
+          top: 1px;
+        }
+      </style>
+    `;
+  }
+
+  private async _fetchData(): Promise<ShoppingListItem[]> {
+    if (this.hass) {
+      this._items = await fetchItems(this.hass);
+      this._items.reverse();
+    }
+
+    return this._items!;

```

Comment:
Where is this used?

Reply:
The return? TypeScript complained if I didn't return a Promise from the async function

Reply:
`this._items` is not a promise, so you're not returning a promise. If TypeScript complains, you might have the type declaration of the function wrong